contract_isp_invoice/models/contract.py

```diff
@@ -0,0 +1,535 @@
+                        account_analytic_journal journal ON
+                        (line.journal_id = journal.id)
+                        WHERE account_id = %s
+                            AND line.id IN %s AND journal.type = %s AND
+                            to_invoice IS NOT NULL
+                            GROUP BY product_id, user_id, to_invoice,"""
+                            """ product_uom_id""", (account.id, tuple(self.ids),
+                                                journal_type))
+
+                for product_id, user_id, factor_id, total_price, qty, uom in\
+                self._cr.fetchall():
+                    context2.update({'uom': uom})
+
+                    if data.get('product'):
+                        # force product, use its public price
+                        product_id = data['product'][0]
+                        unit_price = self._get_invoice_price(account,
+                                                             product_id,
+                                                             user_id, qty)
+                    #  elif journal_type == 'general' and product_id:
+                    #  timesheets, use sale price
+                    #  unit_price = self._get_invoice_price(cr, uid,
+                    #  account, product_id, user_id, qty, context2)
+                    else:
+                        #  expenses, using price from amount field
+                        unit_price = qty and total_price * -1.0 / qty or 0.0
+
+                    factor = invoice_factor_obj.browse(factor_id)
+                    # factor_name = factor.customer_name and line_name +
+                    # ' - ' + factor.customer_name or line_name
+                    factor_name = ''
+                    if data.get('factor_name', False):
+                        factor_name = factor.customer_name
+
+                    curr_line = {
+                        'price_unit': unit_price,
+                        'quantity': qty,
+                        'product_id': product_id or False,
+                        'discount': factor.factor,
+                        'invoice_id': last_invoice.id,
+                        'name': factor_name,
+                        'uos_id': uom,
+                        'account_analytic_id': account.id,
+                    }
+                    product = product_obj.browse(product_id)
+                    if product:
+                        factor_name = data.get('product_name', '') and \
+                            product.name_get()
+                        if factor.customer_name and data.get('factor_name',
+                                                             False):
+                            factor_name += ' - ' + factor.customer_name
+
+                        general_account = product.property_account_income or \
+                            product.categ_id.property_account_income_categ
+                        if not general_account:
```

Comment:
checks and warning should be done at the beginning of the method, not in the middle